Doc/library/__main__.rst

```diff
@@ -1,25 +1,212 @@
+`setuptools. <https://setuptools.readthedocs.io/en/latest/>`__
+
+*main* functions are often used to create command line tools by specifying them
+as entry points for console scripts.  When this is done, pip inserts the
+function call into a template that looks like this::
+
+   #!/path/to/python3
+   # -*- coding: utf-8 -*-
+   import re
+   import sys
+   from package.__main__ import main
+   if __name__ == '__main__':
+       sys.argv[0] = re.sub(r'(-script\.pyw|\.exe)?$', '', sys.argv[0])
+       sys.exit(main())
+
+Notice the **last line.** The call to *main* is wrapped in :func:`sys.exit`.
+When *main* is the entry point of a console_script, the expectation is that
+your function will return some value acceptable as an input to
+:func:`sys.exit`; typically, an integer or ``None`` (which is implicitly returned
+if your function does not have a return statement).
+
+By proactively following this convention ourselves, our module will have the
+same behavior when run directly (``python3 echo.py``) as it will have if we
+later package it as an console script entry-point in a pip-installable package.
+We can revise the :file:`echo.py` example from earlier to follow this
+convention::
+
+    # echo.py
+    ...
+
+    def main() -> int:  # now, main returns an integer
+        "Echo the string to standard output"
+        echo(shlex.join(sys.argv))
+        return 0
+
+    if __name__ == '__main__':
+        # now, the integer returned from main is passed through to sys.exit
+        sys.exit(main())
+
+
+``__main__.py`` in Python Packages
+----------------------------------
+
+If you are not familiar with Python packages, see section :ref:`tut-packages`.
+Most commonly, the ``__main__.py`` file is used to provide a command line
+interface for a package. Consider the following hypothetical package,
+"bandclass":
+
+.. code-block:: text
+
+   bandclass
+     ├── __init__.py
+     ├── __main__.py
+     ├── parent.py
+     └── student.py
+
+``__main__.py`` will be executed when the package itself is invoked
+directly from the command line using the :option:`-m` flag. For example::
+
+    python3 -m bandclass
+
+This command will cause ``__main__.py`` to run. For more details about the
+:option:`-m` flag, see :mod:`runpy`. How you utilize this mechanism will depend
+on the nature of the package you are writing, but in this hypothetical case, it
+might make sense to allow the teacher to search for students or parents using
+:mod:`argparse`::
+
+    # bandclass/__main__.py
+
+    import argparse
+    import sys
+
+    from .parent import Parents
+    from .student import Students
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--student',
+                        help="lookup a student and print their information")
+    parser.add_argument('--parent',
+                        help="lookup a parent and print their information")
+
+    args = parser.parse_args()
+
+    if args.student and student := Students.find(args.student):
```

Comment:
> if you say --student foo and foo does not exist, then you want a 'student not found' error, not the --help text.

The `else` case is just the case that whatever you are looking for isn't found, regardless of what you are searching for. I'll take out the `print_help` thing, though, because I agree that you don't want help to be printed in there; argparse provides a `-h` flag anyway.

> := is controversial and should be used only when it provides significant value

I understand that it is controversial, but it is a language feature. I definitely agree that the example had some issues, but I think I've fixed some of them while keeping the :=. Walrus operator aside, what do you think about the fixes I made to the example in ceadd31518652e22f6ef3d0b720cad8437dc2bed? Are there still reasons you'd like to see the walrus gone?

Reply:
Yes, two issues:
1. Docs should demonstrate idiomatic uncontroversial code, and walrus is controversial in itself. Using walrus in official docs serves as a recommendation that walrus should in general be used in this context. If you want to make this kind of recommendation, then that should be done in the docs of `argparse`. The example should be focused on what we are discussing here, which is that *some* parsing of argv is done. Better to keep it short and dull, than showing off Python features.
2. Your code still has behavioural surprises, e.g. `--parent Y --student X` will silently ignore the `--parent` arg. This glitch would not scale to bigger programs, hence I would not say the pattern qualifies as idiomatic Python.

Maybe you can further reduce the cognitive weight of your example by changing the example to some well-known trivial algorithm that needs no presentation, like:
```
import sys
from .mul import product
factors = [float(arg) for arg in sys.argv[1:]]
print(f'The product is {product(factors)}')
```
In my opinion, the most important line to communicate is the relative import. If you add proper error checking with argparse, the example will grow by x2, which shifts the reader's focus.

Reply:
> In my opinion, the most important line to communicate is the relative import.

agree!

Reply:
Addressed in 7c6b451c35787dd62f2431adebf548f9beb6b9d1